File: intcode.py
#!/usr/bin/env python3

import logging

from enum import Enum
from typing import Any, Dict, List, Optional, Sequence, Tuple


Data = List[int]

Address = int


class AddressMode(Enum):
    POSITION = 0
    IMMEDIATE = 1
    RELATIVE = 2

    def __str__(self) -> str:
        return self.name


class Opcode(Enum):
    UNKNOWN = 0
    ADD = 1
    MULTIPLY = 2
    INPUT = 3
    OUTPUT = 4
    JCC = 5
    JNC = 6
    LT = 7
    EQ = 8
    RELBASE = 9
    HALT = 99


OPCODE_SIGNATURES: Dict[Opcode, Tuple[str, List[str]]] = {
    Opcode.UNKNOWN: ("unknown", []),
    Opcode.ADD: ("add", ["a", "b", "addr"]),
    Opcode.MULTIPLY: ("multiply", ["a", "b", "addr"]),
    Opcode.INPUT: ("input", ["addr"]),
    Opcode.OUTPUT: ("output", ["addr"]),
    Opcode.JCC: ("jump-if-true", ["v", "addr"]),
    Opcode.JNC: ("jump-if-not-true", ["v", "addr"]),
    Opcode.LT: ("less-than", ["a", "b", "addr"]),
    Opcode.EQ: ("equal", ["a", "b", "addr"]),
    Opcode.RELBASE: ("adjust-rel-base", ["offset"]),
    Opcode.HALT: ("halt", []),
}


logger = logging.getLogger(__name__)


def decode(s: str) -> Data:
    return [int(opcode) for opcode in s.split(",")]


class IntcodeVM:
    def __init__(
        self, data: Optional[Data] = None, stdin: Optional[Data] = None
    ) -> None:
        self.handlers = {
            Opcode.ADD: self._op_add,
            Opcode.MULTIPLY: self._op_multiply,
            Opcode.INPUT: self._op_input,
            Opcode.OUTPUT: self._op_output,
            Opcode.JCC: self._op_jump_if_true,
            Opcode.JNC: self._op_jump_if_not_true,
            Opcode.LT: self._op_less_than,
            Opcode.EQ: self._op_equal,
            Opcode.RELBASE: self._op_relbase,
        }
        self.reset(data, stdin)

    def reset(self, data: Optional[Data], stdin: Optional[Data] = None) -> None:
        # program counter
        self.reg_pc = Address(0)

        # dedicated register for the relative offset base
        self.reg_rel_base = Address(0)

        # vm memory
        self.data = data.copy() if data else []

        # input buffer
        self.stdin: Data = stdin.copy() if stdin else []

        # output buffer
        self.stdout: Data = []

        # current instruction opcode
        self.opcode: Opcode = Opcode.UNKNOWN

        # current instruction parameter address modes
        self.param_modes: List[AddressMode] = [AddressMode.POSITION] * 4

    def _trace(
        self,
    ) -> None:
        sig = OPCODE_SIGNATURES.get(self.opcode)
        if sig is None:
            sig = OPCODE_SIGNATURES[Opcode.UNKNOWN]
        opcode = sig[0]
        params = sig[1]

        def encode_param(i: int) -> str:
            addr = self.reg_pc + 1 + i
            mode = self.param_modes[i]
            name = params[i]
            v = self.data[addr]
            if mode == AddressMode.POSITION:
                return f"{name}={v}"
            if mode == AddressMode.IMMEDIATE:
                return f"{name}=#{v}"
            if mode == AddressMode.RELATIVE:
                return f"{name}={v}+${self.reg_rel_base}"
            raise RuntimeError(f"Unrecognized address mode: {mode}")

        params_str = ", ".join([encode_param(i) for i in range(len(params))])
        raw_str = ",".join(
            str(i) for i in self.data[self.reg_pc : self.reg_pc + len(params) + 1]
        )
        logger.debug(f"[{self.reg_pc:04}] {opcode}({params_str})  # {raw_str}")

    @property
    def halted(self) -> bool:
        return self.opcode == Opcode.HALT

    def step(self) -> bool:
        """
        Fetch, decode and execute a single instruction.
        """
        if self.halted:
            raise RuntimeError("Halted")

        # fetch instruction
        v = self.read(self.reg_pc)

        # decode opcode and parameter address modes
        self.opcode = Opcode(v % 100)
        v = int(v / 100)
        for i in range(len(self.param_modes)):
            self.param_modes[i] = AddressMode(v % 10)
            v = int(v / 10)
        self._trace()

        # dispatch
        if self.opcode == Opcode.HALT:
            return False
        if self.opcode not in self.handlers:
            raise RuntimeError(f"Unrecognized opcode: {self.opcode}")
        self.handlers[self.opcode]()
        return True

    def run(self) -> None:
        """
        Call step repeatedly until the VM is halted.
        """
        while self.step():
            pass

    def io_push(self, v: int) -> None:
        """
        Push a new value to the VM's input buffer.
        """
        self.stdin.append(v)

    def io_pop(self) -> int:
        """
        Pop a value from the VM's output buffer.
        """
        v = self.stdout[0]
        self.stdout = self.stdout[1:]
        return v

    def effective_addr(self, addr: int, mode: AddressMode) -> int:
        if mode == AddressMode.IMMEDIATE:
            return addr
        if mode == AddressMode.POSITION:
            return self.data[addr]
        if mode == AddressMode.RELATIVE:
            return self.data[addr] + self.reg_rel_base
        raise RuntimeError(f"Unrecognized address mode: {mode}")

    def grow_mem(self, addr: int) -> int:
        old_size = len(self.data)
        if addr < old_size:
            return old_size
        new_size = 64
        while new_size < addr:
            new_size *= 2
        if new_size > 4096:
            raise RuntimeError("Too much memory allocated")
        new_data = [0] * new_size
        for i in range(old_size):
            new_data[i] = self.data[i]
        self.data = new_data
        logger.debug(f"  grow_mem(addr={addr}) -> {new_size}")
        return new_size

    def read(self, addr: int, mode=AddressMode.IMMEDIATE) -> int:
        if addr != AddressMode.IMMEDIATE:
            addr = self.effective_addr(addr, mode)
        assert addr >= 0, f"Invalid address: {addr}"
        if addr >= len(self.data):
            v = 0
        else:
            v = self.data[addr]
        logger.debug(f"  read(addr={addr}) -> {v}")
        return v

    def write(self, addr: int, v: int) -> None:
        self.grow_mem(addr)
        self.data[addr] = v
        logger.debug(f"  write(addr={addr}, v={v})")

    def _op_add(self) -> None:
        """
        Opcode 1 adds together numbers read from two positions and stores the result in a third
        position. The three integers immediately after the opcode tell you these three positions
         - the first two indicate the positions from which you should read the input values, and the
        third indicates the position at which the output should be stored.
        """
        a = self.read(self.reg_pc + 1, self.param_modes[0])
        b = self.read(self.reg_pc + 2, self.param_modes[1])
        addr = self.effective_addr(self.reg_pc + 3, self.param_modes[2])
        v = a + b
        self.write(addr, v)
        self.reg_pc += 4

    def _op_multiply(self) -> None:
        """
        Opcode 2 works exactly like opcode 1, except it multiplies the two inputs instead of adding
        them. Again, the three integers after the opcode indicate where the inputs and outputs are,
        not their values.
        """
        a = self.read(self.reg_pc + 1, self.param_modes[0])
        b = self.read(self.reg_pc + 2, self.param_modes[1])
        addr = self.effective_addr(self.reg_pc + 3, self.param_modes[2])
        v = a * b
        self.write(addr, v)
        self.reg_pc += 4

    def _op_input(self) -> None:
        if self.stdin:
            v = self.stdin[0]
            self.stdin = self.stdin[1:]
        else:
            v = int(input(f"[{self.reg_pc:04}] Enter input: "))
        addr = self.effective_addr(self.reg_pc + 1, self.param_modes[0])
        self.write(addr, v)
        self.reg_pc += 2

    def _op_output(self) -> None:
        v = self.read(self.reg_pc + 1, self.param_modes[0])
        self.stdout.append(v)
        self.reg_pc += 2

    def _op_jump_if_true(self) -> None:
        """
        Opcode 5 is jump-if-true: if the first parameter is non-zero, it sets the instruction
        pointer to the value from the second parameter. Otherwise, it does nothing.
        """
        a = self.read(self.reg_pc + 1, self.param_modes[0])
        addr = self.read(self.reg_pc + 2, self.param_modes[1])
        v = a != 0
        if v:
            self.reg_pc = addr
        else:
            self.reg_pc += 3

    def _op_jump_if_not_true(self) -> None:
        """
        Opcode 6 is jump-if-false: if the first parameter is zero, it sets the instruction pointer
        to the value from the second parameter. Otherwise, it does nothing.
        """
        a = self.read(self.reg_pc + 1, self.param_modes[0])
        addr = self.read(self.reg_pc + 2, self.param_modes[1])
        v = a == 0
        if v:
            self.reg_pc = addr
        else:
            self.reg_pc += 3

    def _op_less_than(self) -> None:
        """
        Opcode 7 is less than: if the first parameter is less than the second parameter, it stores 1
        in the position given by the third parameter. Otherwise, it stores 0.
        """
        a = self.read(self.reg_pc + 1, self.param_modes[0])
        b = self.read(self.reg_pc + 2, self.param_modes[1])
        addr = self.effective_addr(self.reg_pc + 3, self.param_modes[2])
        v = int(a < b)
        self.write(addr, v)
        self.reg_pc += 4

    def _op_equal(self) -> None:
        """
        Opcode 8 is equals: if the first parameter is equal to the second parameter, it stores 1 in
        the position given by the third parameter. Otherwise, it stores 0.
        """
        a = self.read(self.reg_pc + 1, self.param_modes[0])
        b = self.read(self.reg_pc + 2, self.param_modes[1])
        addr = self.effective_addr(self.reg_pc + 3, self.param_modes[2])
        v = int(a == b)
        self.write(addr, v)
        self.reg_pc += 4

    def _op_relbase(self) -> None:
        """
        Opcode 9 adjusts the relative base by the value of its only parameter. The relative base
        increases (or decreases, if the value is negative) by the value of the parameter.
        """
        a = self.read(self.reg_pc + 1, self.param_modes[0])
        self.reg_rel_base += a
        self.reg_pc += 2


def run(data: Data, stdin: Optional[Data] = None) -> Data:
    vm = IntcodeVM(data=data, stdin=stdin)
    vm.run()
    return vm.stdout


def load_and_run(file: str, stdin: Optional[Data] = None) -> Data:
    with open(file, "r") as fp:
        data = decode(fp.readline())
    return run(data, stdin=stdin)

File: go/intcode/intcode.go
package intcode

import (
	"bytes"
	"errors"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"strconv"
	"strings"
)

var (
	ErrIllegal  = errors.New("illegal")
	ErrHalted   = errors.New("halted")
	ErrNoInput  = errors.New("no input")
	ErrNoOutput = errors.New("no output")
)

// Data is a slice of program and data memory for the VM.
type Data []int

// DecodeData reads Data from a file and decodes it from a comma-seperated
// string format.
func DecodeData(r io.Reader) (data Data, err error) {
	var b []byte
	var v int
	b, err = ioutil.ReadAll(r)
	if err != nil {
		return
	}
	parts := bytes.Split(b, []byte{','})
	data = make([]int, len(parts))
	for i := 0; i < len(parts); i++ {
		v, err = strconv.Atoi(strings.TrimRight(string(parts[i]), "\n"))
		if err != nil {
			return
		}
		data[i] = v
	}
	return
}

// OpenData reads all data from a file.
func OpenData(name string) (data Data, err error) {
	var f *os.File
	f, err = os.Open(name)
	if err != nil {
		return
	}
	defer f.Close()
	return DecodeData(f)
}

// Opcode is the ID of a function that the VM can perform.
type Opcode int

const (
	OpcodeIllegal Opcode = iota
	OpcodeAdd
	OpcodeMultiply
	OpcodeInput
	OpcodeOutput
	OpcodeJCC
	OpcodeJNC
	OpcodeLessThan
	OpcodeEqual
	OpcodeRelBase
	OpcodeHalt = 99
)

type opcodeSignature struct {
	name   string
	params []string
}

var opcodeSignatures = map[Opcode]opcodeSignature{
	OpcodeAdd:      {"add", []string{"a", "b", "addr"}},
	OpcodeMultiply: {"multiple", []string{"a", "b", "addr"}},
	OpcodeInput:    {"input", []string{"addr"}},
	OpcodeOutput:   {"output", []string{"addr"}},
	OpcodeJCC:      {"jump-if-true", []string{"v", "addr"}},
	OpcodeJNC:      {"jump-if-false", []string{"v", "addr"}},
	OpcodeLessThan: {"less-than", []string{"a", "b", "addr"}},
	OpcodeEqual:    {"equal", []string{"a", "b", "addr"}},
	OpcodeRelBase:  {"adjust-rel-base", []string{"offset"}},
	OpcodeHalt:     {"halt", []string{}},
}

type opcodeHandler func() error

// AddressMode describes how an Address should be dereferenced by the VM.
type AddressMode int

const (
	// AddressModePosition reads the memory address of a value.
	AddressModePosition AddressMode = iota

	// AddressModeImmediate reads a value directly.
	AddressModeImmediate

	// AddressModeRelative reads the memory address of a value relative to the
	// value of the Relative Base Register.
	AddressModeRelative
)

// Address points to a location in VM mempory.
type Address int

// Instruction is an integer encoding of an opcode and the address mode of each
// of its parameters.
type Instruction int

// Opcode returns the Opcode of the instruction.
func (c Instruction) Opcode() Opcode {
	return Opcode(c) % 100
}

// Modes returns the address mode of each paramter in the instruction.
func (c Instruction) Modes() [4]AddressMode {
	modes := [4]AddressMode{0, 0, 0, 0}
	v := AddressMode(c) / 100
	for i := 0; i < len(modes); i++ {
		modes[i] = v % 10
		v /= 10
	}
	return modes
}

// A VirtualMachine runs IntCode programs.
type VirtualMachine interface {
	// Step will fetch, decode and execute a single instruction
	Step() error

	// Run until the program halts
	Run() error

	// MemGet reads a value from VM memory
	MemGet(addr Address, mode AddressMode) int

	// MemSet writes a value to VM memory
	MemSet(addr Address, v int)

	// IOPush sends a value to the VM input buffer
	IOPush(v int, block bool) (err error)

	// IOPop reads a value from the VM output buffer
	IOPop(block bool) (v int, err error)
}

type virtualMachine struct {
	data        Data
	handlers    []opcodeHandler
	stdin       []int
	stdout      []int
	opcode      Opcode
	modes       [4]AddressMode
	regPC       Address
	regRelBase  Address
	traceWriter io.Writer
}

// New returns a new VirtualMachine.
func New(data Data) VirtualMachine {
	v := &virtualMachine{
		data:   data,
		stdin:  make([]int, 0, 64),
		stdout: make([]int, 0, 64),
	}
	v.handlers = []opcodeHandler{
		v.opIllegal,
		v.opAdd,
		v.opMultiply,
		v.opInput,
		v.opOutput,
		v.opJumpIfTrue,
		v.opJumpIfNotTrue,
		v.opLessThan,
		v.opEqual,
		v.opRelBase,
	}
	if os.Getenv("LOGLEVEL") == "DEBUG" {
		v.traceWriter = os.Stderr
	}
	return v
}

func (c *virtualMachine) tracef(format string, a ...interface{}) {
	if c.traceWriter == nil {
		return
	}
	fmt.Fprintf(c.traceWriter, format, a...)
}

func (c *virtualMachine) traceStep() {
	if c.traceWriter == nil {
		return
	}
	sig := opcodeSignatures[c.opcode]
	fmt.Fprintf(c.traceWriter, "[%04d] %s(", c.regPC, sig.name)
	for i := 0; i < len(sig.params); i++ {
		if i > 0 {
			fmt.Fprintf(c.traceWriter, ", ")
		}
		v := c.data[c.regPC+Address(1+i)]
		switch c.modes[i] {
		case AddressModePosition:
			fmt.Fprintf(c.traceWriter, "%s=%d", sig.params[i], v)
		case AddressModeImmediate:
			fmt.Fprintf(c.traceWriter, "%s=#%d", sig.params[i], v)
		case AddressModeRelative:
			fmt.Fprintf(c.traceWriter, "%s=%d+$%d", sig.params[i], v, c.regRelBase)
		}
	}
	fmt.Fprintf(c.traceWriter, ")  // ")
	for i := 0; i < len(sig.params)+1; i++ {
		if i > 0 {
			fmt.Fprintf(c.traceWriter, ", ")
		}
		fmt.Fprintf(c.traceWriter, "%d", c.data[c.regPC+Address(i)])
	}
	fmt.Fprintf(c.traceWriter, "\n")
}

func (c *virtualMachine) Step() error {
	// fetch and decode
	v := Instruction(c.MemGet(c.regPC, AddressModeImmediate))
	c.opcode = v.Opcode()
	c.modes = v.Modes()
	c.traceStep()

	// dispatch
	if c.opcode == OpcodeHalt {
		return ErrHalted
	}
	return c.handlers[c.opcode]()
}

func (c *virtualMachine) Run() error {
	for {
		if err := c.Step(); err != nil {
			if err == ErrHalted {
				return nil
			}
			return err
		}
	}
}

func (c *virtualMachine) MemGet(addr Address, mode AddressMode) (v int) {
	eaddr := c.effectiveAddr(addr, mode)
	if int(eaddr) >= len(c.data) {
		v = 0
	} else {
		v = c.data[eaddr]
	}
	c.tracef("  memget(%04d, %d) -> %v\n", addr, mode, v)
	return
}

func (c *virtualMachine) MemSet(addr Address, v int) {
	if int(addr) >= len(c.data) {
		// grow mem
		sz := 64
		for sz < int(addr)+1 {
			sz *= 2
		}
		data := make(Data, sz)
		copy(data, c.data)
		c.data = data
	}

	c.data[addr] = v
	c.tracef("  memset(%04d, %d)\n", addr, v)
}

func (c *virtualMachine) IOPush(v int, block bool) (err error) {
	c.tracef("  iopush(%d)\n", v)
	c.stdin = append(c.stdin, v)
	for block && len(c.stdin) > 0 {
		err = c.Step()
		if err != nil {
			return
		}
	}
	return nil
}

func (c *virtualMachine) IOPop(block bool) (v int, err error) {
	for block && len(c.stdout) == 0 {
		err = c.Step()
		if err != nil {
			c.tracef("  iopop() -> ERROR\n")
			return
		}
	}
	if len(c.stdout) == 0 {
		c.tracef("  iopop() -> ERROR\n")
		return 0, ErrNoOutput
	}
	v = c.stdout[0]
	c.stdout = c.stdout[1:]
	c.tracef("  iopop() -> %d\n", v)
	return
}

func (c *virtualMachine) effectiveAddr(addr Address, mode AddressMode) Address {
	switch mode {
	case AddressModeImmediate:
		return addr
	case AddressModePosition:
		return Address(c.MemGet(addr, AddressModeImmediate))
	case AddressModeRelative:
		return Address(c.MemGet(addr, AddressModeImmediate)) + c.regRelBase
	}
	panic(mode)
}

func (c *virtualMachine) opIllegal() error {
	return ErrIllegal
}

func (c *virtualMachine) opAdd() error {
	a := c.MemGet(c.regPC+1, c.modes[0])
	b := c.MemGet(c.regPC+2, c.modes[1])
	addr := c.effectiveAddr(c.regPC+3, c.modes[2])
	v := a + b
	c.MemSet(addr, v)
	c.regPC += 4
	return nil
}

func (c *virtualMachine) opMultiply() error {
	a := c.MemGet(c.regPC+1, c.modes[0])
	b := c.MemGet(c.regPC+2, c.modes[1])
	addr := c.effectiveAddr(c.regPC+3, c.modes[2])
	v := a * b
	c.MemSet(addr, v)
	c.regPC += 4
	return nil
}

func (c *virtualMachine) opInput() error {
	addr := c.effectiveAddr(c.regPC+1, c.modes[0])
	if len(c.stdin) == 0 {
		return ErrNoInput
	}
	v := c.stdin[0]
	c.stdin = c.stdin[1:]
	c.MemSet(addr, v)
	c.regPC += 2
	return nil
}

func (c *virtualMachine) opOutput() error {
	v := c.MemGet(c.regPC+1, c.modes[0])
	c.stdout = append(c.stdout, v)
	c.regPC += 2
	return nil
}

func (c *virtualMachine) opJumpIfTrue() error {
	v := c.MemGet(c.regPC+1, c.modes[0])
	addr := Address(c.MemGet(c.regPC+2, c.modes[1]))
	if v != 0 {
		c.regPC = addr
	} else {
		c.regPC += 3
	}
	return nil
}

func (c *virtualMachine) opJumpIfNotTrue() error {
	v := c.MemGet(c.regPC+1, c.modes[0])
	addr := Address(c.MemGet(c.regPC+2, c.modes[1]))
	if v == 0 {
		c.regPC = addr
	} else {
		c.regPC += 3
	}
	return nil
}

func (c *virtualMachine) opLessThan() error {
	a := c.MemGet(c.regPC+1, c.modes[0])
	b := c.MemGet(c.regPC+2, c.modes[1])
	addr := (c.effectiveAddr(c.regPC+3, c.modes[2]))
	v := 0
	if a < b {
		v = 1
	}
	c.MemSet(addr, v)
	c.regPC += 4
	return nil
}

func (c *virtualMachine) opEqual() error {
	a := c.MemGet(c.regPC+1, c.modes[0])
	b := c.MemGet(c.regPC+2, c.modes[1])
	addr := (c.effectiveAddr(c.regPC+3, c.modes[2]))
	v := 0
	if a == b {
		v = 1
	}
	c.MemSet(addr, v)
	c.regPC += 4
	return nil
}

func (c *virtualMachine) opRelBase() error {
	c.regRelBase += Address(c.MemGet(c.regPC+1, c.modes[0]))
	c.regPC += 2
	return nil
}
